### PR TITLE
Update path instructions post install

### PR DESCRIPTION
**Pull request by ChaseAllbee**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: scripts/bundled_installer [hunk 1]
```diff
@@ -50,7 +51,8 @@ function install_ebcli() {
     if [ -f ebcli_installer.py ]; then
         ${PYENV_BIN}/python ./ebcli_installer.py \
             --python-installation ${PYENV_BIN}/python \
-            --virtualenv-executable ${PYENV_BIN}/virtualenv
+            --virtualenv-executable ${PYENV_BIN}/virtualenv \
+            --script-source ${SCRIPT_SOURCE}
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: The responsibility of printing the Python `PATH`-`export` message must not be that of `ebcli_installer.py`. It should be that of the `bundled_installer`.

The way we should deal with printing the Python PATH-export message is as follows:

1. pass a flag to `python_installer`, `--suppress-path-export-message`, so that it can avoid displaying the message
2. print the export message that `python_installer` in `bundled_installer` if and only if Python was actually installed (`python_installer` might have found Python already in PATH and hence skipped installation). For this, you can set a boolean in python_installer.
3. you should ensure that even if `ebcli_installer` fails, the Python export statement is printed. I am concerned here that the error message in red followed by the green + yellow Python PATH-export messages might make the customer believe that everything went alright, which I think, is why I was hesitant to put the PATH-export messages for both Python and the EBCLI together in the first place.

Comment (rahulrajaram) [reply scripts/bundled_installer]: > 2.  ... If there is a way to do this that I'm unaware of then this approach also makes sense.

You can pass arguments to `source python_installer`.

> 3. ... However I think that concern would support the idea of displaying neither export statement.

The concern supports the idea of not making the change at all so that the python_installer continue printing the PATH-export message where it currently is. Note, however, I am not insisting we retain the behaviour.

> 3. ... What usecase would require a user to want to add python to path without also having succeeded the ebcli installer?

That's a fair point. Considering the python_installer is exclusive to the ebcli_installation, we can avoid showing the Python PATH-export statement even if Python was installed.

Comment (ChaseAllbee) [reply scripts/bundled_installer]: Just updated the commit to reflect these changes

File: scripts/ebcli_installer.py [hunk 2]
```diff
@@ -151,6 +151,31 @@ def _exec_cmd(args):
 """
 
 
+INSTALLATION_SUCCESS_MESSAGE_WITH_EXPORT_RECOMMENDATION__NON_WINDOWS__PYTHON_INCLUDED = """
```
Comment (rahulrajaram) [inline scripts/ebcli_installer.py]: We discussed about not touching this file at all.

Comment (ChaseAllbee) [reply scripts/ebcli_installer.py]: Ah looks like that was just an artifact from the previous commit, I'll remove it now

Comment (ChaseAllbee) [reply scripts/ebcli_installer.py]: Updated that file

File: scripts/python_installer [hunk 3]
```diff
@@ -24,7 +24,9 @@ BASH_PROFILE="$HOME/.bash_profile"
 ZSHENV="$HOME/.zshrc"
 COMMAND_LINE_ARGS=$*
 PYTHON_VERSION=${PYTHON_VERSION:="3.7.2"}
+SURPRESS_PATH_EXPORT_MESSAGE=${SURPRESS_PATH_EXPORT_MESSAGE:="false"}
```
Comment (rahulrajaram) [inline scripts/python_installer]: We should be able to booleans, `true` and `false`, rather than the strings, "true" and "false".

File: scripts/python_installer [hunk 4]
```diff
@@ -160,7 +168,7 @@ function verify_http_client_exists() {
         echo_error_message "wget (>= 1.14) or curl (>= 7.18.1) is required to download and install Python."
         exit 1
     elif [ -n "$(wget --version 2>/dev/null)" ]; then
-        return
+-        return
```
Comment (rahulrajaram) [inline scripts/python_installer]: Undo this change.

File: scripts/bundled_installer [hunk 5]
```diff
@@ -55,6 +69,26 @@ function install_ebcli() {
     fi
 }
 
+function print_path_export_instructions__python() {
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Why not `export` the equivalent definitions in `python_installer` and simply use them here?

File: scripts/bundled_installer [hunk 6]
```diff
@@ -31,14 +34,25 @@ function exit_if_return_code_is_non_zero() {
     fi
 }
 
+function echo_with_indentation() {
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Why not `export` the equivalent definitions in `python_installer` and simply use them here?

File: scripts/bundled_installer [hunk 7]
```diff
@@ -10,6 +10,9 @@
 export PYTHON_VERSION="3.7.2"
 export PYENV_ROOT="$HOME/.pyenv"
 export PYENV_BIN="$PYENV_ROOT/versions/$PYTHON_VERSION/bin"
+export SURPRESS_PATH_EXPORT_MESSAGE="true"
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: You don't need to export this statement. Since you are `source`-ing `python_installer`, you can pass this as an argument:

```sh
         source ./python_installer SURPRESS_PATH_EXPORT_MESSAGE=true
```

Also,:
- the value can be a boolean rather than a string
- move this assignment into the `install_python` function defined below to keep its scope limited.

File: scripts/bundled_installer [hunk 8]
```diff
@@ -55,6 +69,26 @@ function install_ebcli() {
...
+function print_export_messages() {
+    if [ ${PYTHON_INSTALLED_PREVIOUSLY} = "false" ]; then
+        print_path_export_instructions__python
+    fi
+}
+
 change_to_scripts_directory
 install_python
 install_ebcli
+print_export_messages
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Missing trailing newline.

File: scripts/bundled_installer [hunk 9]
```diff
@@ -10,6 +10,9 @@
 export PYTHON_VERSION="3.7.2"
 export PYENV_ROOT="$HOME/.pyenv"
 export PYENV_BIN="$PYENV_ROOT/versions/$PYTHON_VERSION/bin"
+export SURPRESS_PATH_EXPORT_MESSAGE="true"
+BASH_PROFILE="$HOME/.bash_profile"
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: You're defining `BASH_PROFILE` but not `ZSHENV`. You'll need both.

File: scripts/python_installer [hunk 10]
```diff
@@ -105,11 +107,17 @@ function install_python() {
     echo_success_message " - Python $PYTHON_VERSION is installed at $PYENV_BIN"
 
     if python_is_not_in_path; then
-        print_path_export_instructions
+        if [[ "$SURPRESS_PATH_EXPORT_MESSAGE" -ne "true" ]]; then
```
Comment (rahulrajaram) [inline scripts/python_installer]: `-e "false"` would have been preferable, but if you use booleans to represent truthiness, you can simplify as:

```sh
if [ $SURPRESS_PATH_EXPORT_MESSAGE ] ; then
...
```

File: scripts/python_installer [hunk 11]
```diff
@@ -105,11 +107,17 @@ function install_python() {
...
+        fi
     else
+        PYTHON_ALREADY_INSTALLED="true"
         echo ""
-        echo_with_indentation " - $PYENV_BIN is already in PATH no further action need be taken."
+        if [[ "$SURPRESS_PATH_EXPORT_MESSAGE" -ne "true" ]]; then
+            echo_with_indentation " - $PYENV_BIN is already in PATH no further action need be taken."
+        fi
     fi
+
```
Comment (rahulrajaram) [inline scripts/python_installer]: Unnecessary newline.

**Comment by rahulrajaram** [comment]
We should extract the functions and static variable names that are common to `python_installer` and `bundled_installer` and put them in a separate file. We can then `source` this file in each of the former files.

File: scripts/python_installer (hunk 10 above)
Comment (ChaseAllbee) [reply scripts/python_installer]: This makes sense, to add some clarity to what it's doing though I may change the variable name from $SURPRESS_PATH_EXPORT_MESSAGE to $SHOW_PATH_EXPORT_MESSAGE because if we want to test for truthiness then the variable should reflect that.

Comment (ChaseAllbee) [reply scripts/python_installer]: Gave this a shot and it looks like this works only without the brackets. So the statement will be if ${SHOW_PATH_EXPORT_MESSAGE} rather than if [ ${SHOW_PATH_EXPORT_MESSAGE} ]

File: scripts/bundled_installer (hunk 5 above)
Comment (ChaseAllbee) [reply scripts/bundled_installer]: Didn't realize you could export functions, that's a much cleaner way to do that. Will change.

File: scripts/python_installer [hunk 12]
```diff
@@ -24,7 +24,9 @@ BASH_PROFILE="$HOME/.bash_profile"
 ZSHENV="$HOME/.zshrc"
 COMMAND_LINE_ARGS=$*
 PYTHON_VERSION=${PYTHON_VERSION:="3.7.2"}
+SHOW_PATH_EXPORT_MESSAGE=${SHOW_PATH_EXPORT_MESSAGE:=true}
 PYENV_ROOT=${PYENV_ROOT:="$HOME/.pyenv"}
+export PYTHON_ALREADY_INSTALLED="false"
```
Comment (rahulrajaram) [inline scripts/python_installer]: Use boolean here.

File: scripts/python_installer [hunk 13]
```diff
@@ -105,10 +107,15 @@ function install_python() {
     echo_success_message " - Python $PYTHON_VERSION is installed at $PYENV_BIN"
 
     if python_is_not_in_path; then
-        print_path_export_instructions
+        if ${SHOW_PATH_EXPORT_MESSAGE} ; then
+            print_path_export_instructions
+        fi
     else
+        PYTHON_ALREADY_INSTALLED="true"
```
Comment (rahulrajaram) [inline scripts/python_installer]: Use boolean here.

File: scripts/bundled_installer [hunk 14]
```diff
@@ -10,6 +10,9 @@
 export PYTHON_VERSION="3.7.2"
 export PYENV_ROOT="$HOME/.pyenv"
 export PYENV_BIN="$PYENV_ROOT/versions/$PYTHON_VERSION/bin"
+BASH_PROFILE="$HOME/.bash_profile"
+ZSHENV="$HOME/.zshrc"
+PYTHON_INSTALLED_PREVIOUSLY="false"
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Use boolean here. Also, this variable should be called `PYTHON_ALREADY_IN_PATH` because it is set to `true` when `python_is_not_in_path` is `false`

File: scripts/bundled_installer [hunk 15]
```diff
@@ -31,14 +34,21 @@ function exit_if_return_code_is_non_zero() {
...
+
 function install_python() {
     echo ""
+    SHOW_PATH_EXPORT_MESSAGE=false
     echo_with_colors_inverted "=============================================="
     echo_with_colors_inverted "I. Installing Python                          "
     echo_with_colors_inverted "=============================================="
     if [ -f python_installer ]; then
-        ./python_installer
+        source ./python_installer ${SHOW_PATH_EXPORT_MESSAGE}
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: I don't think you need the `${}` enclosure

File: scripts/bundled_installer [hunk 16]
```diff
@@ -31,14 +34,21 @@ function exit_if_return_code_is_non_zero() {
     fi
 }
 
+function echo_recommendation_message() {
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: You should be able to get rid of this function just as you did [here](https://github.com/aws/aws-elastic-beanstalk-cli-setup/pull/54#pullrequestreview-282192972)

File: scripts/bundled_installer [hunk 17]
```diff
@@ -31,14 +34,21 @@ function exit_if_return_code_is_non_zero() {
...
+    local message=$1
+    echo_with_indentation "\033[38;5;11m$message\033[0m"
+}
+
 function install_python() {
     echo ""
+    SHOW_PATH_EXPORT_MESSAGE=false
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: You do not need this until you get to the `if` block below. Move this variable definition inside the `if` block.

File: scripts/bundled_installer [hunk 18]
```diff
@@ -31,14 +34,21 @@ function exit_if_return_code_is_non_zero() {
...
     echo_with_colors_inverted "=============================================="
     echo_with_colors_inverted "I. Installing Python                          "
     echo_with_colors_inverted "=============================================="
     if [ -f python_installer ]; then
-        ./python_installer
+        source ./python_installer ${SHOW_PATH_EXPORT_MESSAGE}
         exit_if_return_code_is_non_zero
+        PYTHON_INSTALLED_PREVIOUSLY=${PYTHON_ALREADY_INSTALLED}
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: This appears to be a redundant assignment.

File: scripts/bundled_installer [hunk 19]
```diff
@@ -55,6 +65,13 @@ function install_ebcli() {
     fi
 }
 
+function print_export_messages() {
+    if [ ${PYTHON_INSTALLED_PREVIOUSLY} = "false" ]; then
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Use boolean here.

File: scripts/bundled_installer [hunk 20]
```diff
@@ -55,6 +65,13 @@ function install_ebcli() {
...
+        print_path_export_instructions
+    fi
+}
+
 change_to_scripts_directory
 install_python
 install_ebcli
+print_export_messages
```
Comment (rahulrajaram) [inline scripts/bundled_installer]: Rename to `print_python_path_export_message()`

File: scripts/python_installer [hunk 21]
```diff
@@ -105,10 +107,15 @@ function install_python() {
     echo_success_message " - Python $PYTHON_VERSION is installed at $PYENV_BIN"
 
     if python_is_not_in_path; then
-        print_path_export_instructions
+        if ${SHOW_PATH_EXPORT_MESSAGE} ; then
```
Comment (rahulrajaram) [inline scripts/python_installer]: You can combine the two `if`s into a multi-operand conditional.

Comment (rahulrajaram) [reply scripts/python_installer]: BTW, the default behaviour of this file must be to show rather than suppress. The prerogative to show the Python-`PATH`-export message must be that of the `python_installer`. The only right that an external entity (such as the `bundled_installer`) may be afforded is that of "requesting" suppression.

Given that, we must rename this variable to `SUPPRESS_PATH_EXPORT_MESSAGE` and set its default value to `false along the lines of what we had discussed [here](https://github.com/aws/aws-elastic-beanstalk-cli-setup/pull/54#discussion_r318889122).

Comment (ChaseAllbee) [reply scripts/python_installer]: I will change the variable name to be more descriptive. 

To your point about the multi-operand conditional, I think the way it is above is suitable for out use. If I combine the the two checks into one, then it would be checking for both python not being in path and if I want to suppress the export message. In the case of wanting to call the print_path_export_instructions, this is the correct usage. However, this misses the case where python is not in path but I do want to suppress this message. In that case I want the script to do nothing. The else statement below it only handles the situation where python is in path and we do want to show or hide the message. The solution here is to either add another case to this if or leave it as is. I think adding an elif would be a little more verbose.

File: scripts/python_installer [hunk 22]
```diff
@@ -105,10 +107,15 @@ function install_python() {
...
     if python_is_not_in_path; then
-        print_path_export_instructions
+        if [ ${SUPPRESS_PATH_EXPORT_MESSAGE} = false ] ; then
+            print_path_export_instructions
+        fi
     else
+        PYTHON_ALREADY_IN_PATH=true
         echo ""
-        echo_with_indentation " - $PYENV_BIN is already in PATH no further action need be taken."
+        if [ ${SUPPRESS_PATH_EXPORT_MESSAGE} = false ] ; then
```
Comment (rahulrajaram) [inline scripts/python_installer]: Curious to know why you made this change.

Comment (ChaseAllbee) [reply scripts/python_installer]: I'm essentially trying to not have the python installer say anything regarding PATH unless it's been run. Like above, this is just a check to see if the script should suppress the output of messages regarding PATH.
